pom.xml

```diff
@@ -323,6 +323,12 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>org.powermock</groupId>
```

Comment:
Testcase DescriptorImplTest#testValidateConnectionError() https://github.com/jenkinsci/jira-plugin/blob/07e34405ccfb469c1a1c4530bc9f415cbbfdb3fa/src/test/java/hudson/plugins/jira/DescriptorImplTest.java#L46 made false positive IIUC. 
```
    @Test
    public void testValidateConnectionError() throws Exception {
        when(jiraSession.getMyPermissions()).thenThrow(RestClientException.class);
        when(jiraSite.createSession()).thenReturn(jiraSession);
        FormValidation validation = descriptor.doValidate("user", "http://localhost:8080", "pass", null, null, false, " ", JiraSite.DEFAULT_TIMEOUT);
        assertEquals(validation.kind, FormValidation.Kind.ERROR);
    }
```
The condition passed to descriptor.doValidate() indeed test alternativeURL.
Since mockito does not support mock new operations, I imported powermock to correct this test.

Reply:
Builder pattern seems not helpful with mock, mock ```JiraSite.createSession()``` in this case. Can you explain more about how to mock with builder pattern? Thanks!

Reply:
In order to properly inject `JiraSite` object in `DescriptorImpl` you might define `getJiraSite()` method and mock it later, but it would require to pass all the parameters you have for `JiraSite` constructor. Instead, you define `getJiraSiteBuilder()` returning just new builder, then you configure builder in your impl code and call `build()` to get actual `JiraSite` constructed. Later in tests, you mock `getJiraSiteBuilder()` to return mocked/spy'ed builder, which would return mocked site on `build()` call.

Reply:
I'm not sure whether I implement the builder pattern right or not. It seems ugly 👎
 I'll include this if it satisfies. See https://github.com/gmshake/jira-plugin/pull/1/files